File: mpd/CLI/Complete.swift
// mpd — shell completion candidate emitter.
//
// Single source of truth for both bash and zsh. The shell-side shims under
// `assets/completions/` invoke `mpd --complete <cword> <word0> <word1> ...`
// and feed the lines we print here back to compadd / COMPREPLY.
//
// Why dynamic instead of a static script: mpd's grammar depends on filesystem
// state (project list, runtime list, db list) and on JSON-driven verb sets
// (per-runtime + per-project-type). Hand-writing the same grammar twice and
// keeping it in sync would drift; Swift already knows everything.
//
// Latency budget: each Tab press forks `mpd`. Keep this path lightweight —
// no podman calls, no network, just state-file reads and asset listing.

import Foundation

extension Mpd.Completion {

    /// Entry point invoked by main.swift's --complete short-circuit.
    /// Prints one candidate per line on stdout. Always exits 0 (errors are
    /// silently mapped to "no candidates" — the shell shouldn't see them).
    static func emit(cword: Int, words: [String]) {
        let prefix = words.indices.contains(cword) ? words[cword] : ""
        let candidates = candidates(cword: cword, words: words, prefix: prefix)
        for c in candidates where c.hasPrefix(prefix) {
            print(c)
        }
    }

    // MARK: - Dispatch

    private static func candidates(cword: Int, words: [String], prefix: String) -> [String] {
        // word[0] == "mpd"
        if cword <= 1 {
            return firstTokenCandidates(prefix: prefix)
        }
        guard words.indices.contains(1) else { return [] }
        let first = words[1]

        // Global option that takes a value: complete the value.
        if cword == 2, first.hasPrefix("--") {
            return globalOptionValueCandidates(forOption: first)
        }

        // `mpd list <TAB>` → suggest entity types.
        if cword == 2, first == "list" {
            return ["projects", "runtimes", "services", "dbs"]
        }

        // Verb-first form: word[1] is a verb → second token is project name.
        if cword == 2, projectVerbs.contains(first) {
            // For `mpd create <new-project>`, no name suggestion list applies
            // (any unused name is fine); for the other verbs, only existing
            // projects make sense.
            return first == "create" ? [] : projectNames()
        }
        if cword >= 3 {
            return verbArgCandidates(verb: first)
        }
        return []
    }

    // MARK: - First token: verbs + global flags

    private static func firstTokenCandidates(prefix: String) -> [String] {
        if prefix.hasPrefix("-") {
            return globalFlags
        }
        // Verbs first (project verbs + the standalone `list`), then global flags.
        return Array(projectVerbs).sorted() + ["list"] + globalFlags
    }

    /// Static list of every long flag/option mpd supports. Mirrors @Flag/@Option
    /// in main.swift. Add new entries here when you add new top-level flags.
    private static let globalFlags: [String] = [
        "--setup",
        "--start",
        "--stop",
        "--uninstall",
        "--setup-info",
        "--runtime-create",
        "--runtime-start",
        "--runtime-stop",
        "--runtime-delete",
        "--runtime",
        "--db-create",
        "--db-start",
        "--db-stop",
        "--db-delete",
        "--status",
        "--yes",
        "--tui",
        "--debug",
        "--help",
    ]

    // MARK: - Value of a global option (cword == 2 and first starts with --)

    private static func globalOptionValueCandidates(forOption flag: String) -> [String] {
        switch flag {
        case "--runtime-start", "--runtime-stop", "--runtime-delete", "--runtime":
            return runtimeNames()
        case "--db-start", "--db-stop", "--db-delete":
            return databaseNames()
        case "--db-create":
            // Common engine:version combos. The DB layer accepts engine alone
            // (defaulted version) plus engine:version.
            return ["postgres", "postgres:17", "mariadb", "mariadb:10.11", "mysql", "mysql:8.4"]
        case "--runtime-create":
            // Suggest runtimes that have an asset definition but aren't created yet.
            return uncreatedRuntimeAssetNames()
        default:
            return []
        }
    }

    // MARK: - Verb argument completion (cword >= 3)
    //
    // Public form: `mpd <verb> <project> [args...]`. word[1] = verb,
    // word[2] = project, word[3..] = verb-specific flags/values.

    private static func verbArgCandidates(verb: String) -> [String] {
        switch verb {
        case "create":
            // No --db: project-type knobs live in mpd.env and are set via configure.
            return ["--type=", "--git-repo=", "--git-branch=", "--git-depth=", "--yes"]
        case "configure":
            // Suggest commonly-set keys; the user can pass any MPD_* key.
            return ["MPD_DB=", "MPD_PHP_VERSION=", "MPD_PHP_MOODLE_BEHAT=", "--yes"]
        case "delete":
            return ["--yes"]
        default:
            return []
        }
    }

    // MARK: - State sources

    /// Names of registered projects from `~/.mpd/machines/<n>/projects.json`.
    private static func projectNames() -> [String] {
        Mpd.Runtime.State.loadProjects().projects.map(\.name)
    }

    /// Names of created runtimes (directories under
    /// `~/.mpd/machines/<n>/runtimes/`).
    private static func runtimeNames() -> [String] {
        let dir = Mpd.Runtime.State.runtimesDir
        guard let entries = try? FileManager.default.contentsOfDirectory(atPath: dir) else {
            return []
        }
        return entries.sorted()
    }

    /// Names of DB containers from `~/.mpd/machines/<n>/databases.json`.
    private static func databaseNames() -> [String] {
        Mpd.Runtime.State.loadDatabases().databases.map(\.databaseId)
    }

    /// Names of runtimes that have asset definitions but no state entry yet —
    /// candidates for `--runtime-create`.
    private static func uncreatedRuntimeAssetNames() -> [String] {
        guard let assets = try? Mpd.Core.Assets.path() else { return [] }
        let runtimesAssetDir = "\(assets)/runtimes"
        guard let assetEntries = try? FileManager.default.contentsOfDirectory(atPath: runtimesAssetDir)
        else { return [] }
        let created = Set(runtimeNames())
        return assetEntries.filter { !created.contains($0) }.sorted()
    }
}

File: go/internal/cli/complete.go
package cli

import (
	"fmt"
	"io"
	"os"
	"sort"
	"strings"

	"github.com/mutms/mpd/go/internal/assets"
	"github.com/mutms/mpd/go/internal/service"
	"github.com/mutms/mpd/go/internal/state"
)

// Shell completion candidates for bash and zsh; the shims under
// assets/completions/ call `mpd --complete <cword> <word0> …`.
// Every Tab press forks mpd: keep this path to state-file reads and
// directory listings — no podman calls, no network.

// ProjectVerbs are the tokens accepted as the first positional argument.
// They are reserved as project names, so a project can never collide with
// a verb.
var ProjectVerbs = []string{"delete", "help", "init", "reset", "run", "start", "status", "stop"}

// IsProjectVerb reports whether a token is a project verb.
func IsProjectVerb(token string) bool {
	for _, v := range ProjectVerbs {
		if v == token {
			return true
		}
	}
	return false
}

// GlobalFlags is every long flag mpd accepts. It mirrors the flag set in
// cmd/mpd/main.go — add new entries in both places.
var GlobalFlags = []string{
	"--vm-setup",
	"--vm-upgrade",
	"--vm-start",
	"--vm-stop",
	"--vm-restart",
	"--service-start",
	"--service-stop",
	"--service-uninstall",
	"--service-purge",
	"--db-create",
	"--db-start",
	"--db-stop",
	"--db-delete",
	"--vm-status",
	"--vm-diag",
	"--web",
	"--yes",
	"--debug",
	"--help",
}

// Complete prints one candidate per line. Errors map to "no candidates"
// rather than surfacing: the shell must never see a diagnostic where it
// expects a word list.
func Complete(out io.Writer, cword int, words []string, s state.Store, a assets.Tree) {
	prefix := ""
	if cword >= 0 && cword < len(words) {
		prefix = words[cword]
	}
	for _, c := range candidates(cword, words, s, a) {
		if strings.HasPrefix(c, prefix) {
			fmt.Fprintln(out, c)
		}
	}
}

func candidates(cword int, words []string, s state.Store, a assets.Tree) []string {
	// words[0] is "mpd" itself.
	if cword <= 1 {
		prefix := ""
		if cword < len(words) {
			prefix = words[cword]
		}
		return firstTokenCandidates(prefix)
	}
	if len(words) < 2 {
		return nil
	}
	first := words[1]

	// A global option that takes a value: complete the value.
	if cword == 2 && strings.HasPrefix(first, "--") {
		return optionValues(first, s, a)
	}
	if cword == 2 && (first == "list" || first == "ls") {
		return []string{"projects", "services", "infra", "dbs", "network"}
	}
	// The second token is a project name. `init` takes a new name, so
	// nothing is suggested there.
	if cword == 2 && (IsProjectVerb(first) || first == "rm") {
		if first == "init" {
			return nil
		}
		return projectNames(s)
	}
	if cword >= 3 {
		return verbArgs(first)
	}
	return nil
}

func firstTokenCandidates(prefix string) []string {
	if strings.HasPrefix(prefix, "-") {
		return GlobalFlags
	}
	out := append([]string{}, ProjectVerbs...)
	sort.Strings(out)
	// "ls" and "rm" are aliases for "list" and "delete".
	return append(append(out, "list", "ls", "rm"), GlobalFlags...)
}

func optionValues(flag string, s state.Store, a assets.Tree) []string {
	switch flag {
	case "--db-start", "--db-stop", "--db-delete":
		return databaseNames(s)
	case "--db-create":
		// The DB layer accepts a bare engine or engine:version; offer both.
		return []string{"postgres", "postgres:17", "mariadb", "mariadb:10.11", "mysql", "mysql:8.4"}
	case "--service-start", "--service-stop", "--service-uninstall", "--service-purge":
		return service.Names()
	default:
		return nil
	}
}

// verbArgs completes the flags of `mpd <verb> <project> …`.
func verbArgs(verb string) []string {
	switch verb {
	case "init":
		// No --db here: project-type knobs live in mpd.env, set via start.
		return []string{"--type=", "--yes"}
	case "start":
		// Commonly-set keys; any MPD_* key is accepted.
		return []string{"MPD_DB=", "MPD_PHP_VERSION=", "MPD_MOODLE_BEHAT="}
	case "delete", "rm", "reset":
		return []string{"--yes"}
	case "status":
		return []string{"--json"}
	default:
		return nil
	}
}

func projectNames(s state.Store) []string {
	var names []string
	for _, p := range s.Projects() {
		names = append(names, p.Name)
	}
	return names
}

func databaseNames(s state.Store) []string {
	var names []string
	for _, d := range s.Databases() {
		names = append(names, d.DatabaseID)
	}
	return names
}

// CompleteFromArgs is the `--complete` entry point: it parses the shim's
// raw argument list and emits candidates. A malformed cword yields no
// candidates rather than an error.
func CompleteFromArgs(out io.Writer, args []string, s state.Store, a assets.Tree) {
	if len(args) == 0 {
		return
	}
	cword := 0
	if _, err := fmt.Sscanf(args[0], "%d", &cword); err != nil {
		return
	}
	Complete(out, cword, args[1:], s, a)
}

// ExitCompletion ends the process cleanly after emitting candidates.
// Completion must always exit 0: a non-zero status makes some shells beep
// or print the command's stderr into the candidate list.
func ExitCompletion() { os.Exit(0) }
